OCSADV-253: End-to-end implementation of ITC service for GMOS imaging

fnussber opened the pull request:
This PR provides a basic but working end-to-end implementation of the ITC service for GMOS (imaging only) without any bells and whistles. Improvements like client-side caching of calculation results etc. are planned for upcoming PRs.

Unfortunately this PR is a bit biggish because I had to rename/refactor a few things to get the TRCP service to work which rippled through to a lot of the files.

Main relevant changes are the inclusion of the ITC bundles into the spdb application and additions to the ```ItcTable``` and ```ItcTableModel``` classes to perform service calls and handle the results. Future refinements will probably see some of the ITC services related tasks (like creation of input parameter objects) being extracted to some utility classes.

Here it would also be nice to know if there is a way we can already anticipate the usage of ITC in PIT? However since PIT uses its own model I am afraid that the translation of instrument configurations into ITC parameter objects would have to be entirely PIT specific(?).

I also have two questions:

* Do we want to set the Serial Version UIDs for service objects manually?
* Is it correct to use the peer obtained by ```ObservingPeer``` for the service calls?

cquiroz commented:
PIT has its own limited model of instruments. Just enough to make proposals, it is even possible that there is not enough information on the model to do ITC calculations, depending on how finely detailed parameters you need.
Eventually there will be a new common model but that's probably way down the line

swalker2m commented:
Just a note on the future caching.  The `edu.gemini.spModel.obs.SPObsCache` is where we cache time consuming calculations that are tied to an observation.  It is automatically cleaned when something is edited so it seems like a natural fit.

cquiroz commented on bundle/edu.gemini.itc.shared/src/main/Java/edu/gemini/itc/shared/ObservationDetails.java:
```diff
@@ -1,9 +1,13 @@
  * Container for observation detail parameters.
  */
-public final class ObservationDetails {
+public final class ObservationDetails implements Serializable {
```
Do they need to be serializables as they are now part of the Obs tree?

swalker2m replied:
If it is passed over trpc then yeah it has to be `Serializable`.

fnussber replied:
Yes, this are the TOs passed around using TRPC, so they all have to be serializable.

cquiroz commented on bundle/edu.gemini.itc/src/main/scala/edu/gemini/itc/service/ItcServiceImpl.scala:
```diff
@@ -21,7 +22,9 @@ class ItcServiceImpl extends ItcService {
     // TODO: for now in most cases where a validation problem should be reported to the user the ITC code throws an exception instead
-    case e: Throwable => ItcResult.forException(e)
+    case e: Throwable =>
+      e.printStackTrace()
```
`printStackTrace`?

fnussber replied:
Me being lazy, sorry. But this will go away soon (once the error messages are displayed in a more useful way in the client).

cquiroz commented on bundle/jsky.app.ot/src/main/scala/jsky/app/ot/editor/seq/ItcTable.scala:
```diff
@@ -55,29 +67,78 @@ trait ItcTable extends Table {
+    val cond = new ObservingConditions(qual.getImageQuality, qual.getCloudCover, qual.getWaterVapor, qual.getSkyBackground, 1.5)
+
+    // get the instrument configuration
+    val filter    = c.config.getItemValue(new ItemKey("instrument:filter")).asInstanceOf[GmosCommonType.Filter]
```
I wish there would be a safer way to do this, it seems possible this could throw NPE or CCE I think it is more of a model's problem rather than ITC though

fnussber replied:
Agreed, this ain't pretty. I get ```Object```s back, and - depending on the instrument - I know their types, there's not much else I can do, I think.

tpolecat replied:
If you're using this API a lot it might make sense to add some bedazzlement to force a checked cast, so you get an early failure rather than a mistyped value getting passed around.

```scala
implicit class ConfigOps(c: Config) {

  def get[A](key: ItemKey(implicit ev: ClassTag[A]): ClassCastException \/ A =
    \/.fromTryCatch(ev.runtimeClass.cast(a).asInstanceOf[A]).leftMap {
      case cce: ClassCastException => cce
      case ex => throw ex
    }

  def unsafeGet[A: ClassTag](key: ItemKey(implicit ev: ClassTag[A]): A =
    get[A](key).fold(throw _, identity)

}

// and then the cast is hidden
c.config.unsafeGet[GmosCommonType.Filter](new ItemKey("instrument:filter"))
```

fnussber replied:
What is the advantage of using ```Class.cast()``` over ```asInstanceOf[]```? Both seem to check the types and throw a CCE during runtime immediately(?).

tpolecat replied:
Sorry, I worded that text very poorly. My point was just that **if** you abstract out the `get` operation then you end up casting to a type variable, which is unsound, and you can use a `ClassTag` to get around this problem.

Here is an example of what I mean:

```scala
scala> def hmm[A](a: Any) = \/.fromTryCatch(a.asInstanceOf[A])
hmm: [A](a: Any)scalaz.\/[Throwable,A]

scala> hmm[Int]("hello")
res15: scalaz.\/[Throwable,Int] = \/-(hello) // <-- WAT .. this is mistyped

scala> res15.toOption
res16: Option[Int] = Some(hello) // <-- still mistyped

scala> res16.map(_ + 1) // finally an operation that catches us
java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
  at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:105)
  at $anonfun$1.apply(<console>:18)
  at scala.Option.map(Option.scala:146)
  ... 43 elided
```

As opposed to

```scala
scala> def hmm[A](a: Any)(implicit ev: ClassTag[A]) = \/.fromTryCatch(ev.runtimeClass.cast(a).asInstanceOf[A])
hmm: [A](a: Any)(implicit ev: scala.reflect.ClassTag[A])scalaz.\/[Throwable,A]

scala> hmm[Int]("hello")
res19: scalaz.\/[Throwable,Int] = -\/(java.lang.ClassCastException: Cannot cast java.lang.String to int)
```

fnussber replied:
Ok, got you. Thx for clarifying.

cquiroz commented on bundle/jsky.app.ot/src/main/scala/jsky/app/ot/editor/seq/ItcTableModel.scala:
```diff
@@ -26,65 +32,114 @@ object ItcTableModel {
+  override def getValueAt(row: Int, col: Int): Object = col match {
+    case c if c <  headers.size             => header(col).value(uniqueSteps(row), res(row))
+    case c if c >= headers.size + keys.size => result(col).value(uniqueSteps(row), res(row))
+    case c                                  => uniqueSteps(row).config.getItemValue(key(col))
```
case _

cquiroz commented on bundle/jsky.app.ot/src/main/scala/jsky/app/ot/editor/seq/ItcTableModel.scala:
```diff
@@ -26,65 +32,114 @@ object ItcTableModel {
+      futureResult  <- f.value                // unwrap future
+      serviceResult <- futureResult.toOption  // unwrap try
+      calcResult    <- serviceResult.toOption // unwrap validation
+    } yield calcResult
```
Does this work because all of them become `Option` right? I wish scala would let you do a for over different types of monads

swalker2m replied:
In general arbitrary monads don't compose.  For example if you have `m1: => Option[A]` and `m2: => Failure \/ B` what would it mean to write:

````scala
for {
  a <- m1
  b <- m2
} yield fun(a,b)
````
What's the result type?  What if `m1` produces `None`, etc.

This looks like it might be a case for monad transformers though.  If we have a task that contains a disjunction `\/` it could be modeled as `EitherT[Task, ItcFailure, A]`.

cquiroz commented on bundle/jsky.app.ot/src/main/scala/jsky/app/ot/editor/seq/ItcTableModel.scala:
```diff
@@ -26,65 +32,114 @@ object ItcTableModel {
+  // TODO: display errors/validation messages in an appropriate way in the UI, for now also print them to console
+  protected def messages(f: Future[ItcService.Result]): String =
+    f.value.fold("Calculating...") {
+      case Failure(t) => t <| (_.printStackTrace()) |> (_.getMessage)  // "Look mummy, there's a spaceship up in the sky!"
```
:smile:

cquiroz commented:
Looks good to me, many package changes and some code updates
I'm not looking forward to another `sbt reload` :+1:

swalker2m commented on bundle/edu.gemini.itc.shared/src/main/scala/edu/gemini/itc/shared/ItcService.scala:
```diff
@@ -53,4 +59,10 @@ object ItcService {
+  def calculate(kc: KeyChain, peer: Peer, source: SourceDefinition, obs: ObservationDetails, cond: ObservingConditions, tele: TelescopeDetails, ins: InstrumentDetails): Future[Result] =
+    TrpcClient(peer).withKeyChain(kc) future { r =>
+      r[ItcService].calculate(source, obs, cond, tele, ins)
+    }
```
Perhaps we don't need the keychain?  Do we do some kind of permission checking on the server side?  It seems like this is all public information that you can get via the web service w/o an account or anything anyway?

fnussber replied:
That's a good point. Indeed the ITC service is public, so I actually don't need any permissions here.

swalker2m commented:
> Do we want to set the Serial Version UIDs for service objects manually?

If you do that then you're saying that you'll manage any future incompatibilities yourself.  As long as we're using trpc, I think it is better to just let them be incompatible when there are differences and let it break.  The OT and ODB pretty much have to be released together anyway and we check the versions of each already to see whether they are compatible.

swalker2m commented:
> Is it correct to use the peer obtained by ObservingPeer for the service calls?

The observing peer won't be defined for PIs. As Carlos was suggesting, perhaps it should call the service at both sites and use the first answer it gets back.

fnussber commented:
Would it make sense to call the service at the site of the ODB you're connected to? But of course this one could be the one that's further away from you. Calling it at both sites seems a bit of an overkill, but I guess it would make sense.

cquiroz commented:
Actually we could do DNS magic for this, a global DNS entry like itc.gemini.edu and internally it would send you to the closest one. It would make the code simpler but I don't know if our DNS can do that

tpolecat commented:
The OT could have a pinger that figures out which peer is closest to you, I guess. But for now I think hitting both and using the first answer makes sense. If it ends up being a performance issue we can revisit later.

fnussber commented:
I implemented some of the review findings. I will deal with using the appropriate "closest" server for the service calls in a later PR. I also added the web module to the spdb server so we can start using it and see if there are any problems with it being part of the spdb. If nobody has additional comments I am going to merge this tonight or tomorrow morning.
